src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,291 @@
+
+        protected abstract TStatementSyntax CreateUsingStatement(TLocalDeclarationSyntax declarationStatement, SyntaxTriviaList sameLineTrivia, SyntaxList<TStatementSyntax> statementsToSurround);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(document, span, context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    FeaturesResources.Introduce_using_statement,
+                    cancellationToken => IntroduceUsingStatementAsync(document, span, cancellationToken)));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax =
+                root.FindNode(selection)?.GetAncestor<TLocalDeclarationSyntax>()
+                ?? root.FindTokenOnLeftOfPosition(selection.End).GetAncestor<TLocalDeclarationSyntax>();
+
+            if (declarationSyntax is null)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var operation = semanticModel.GetOperation(declarationSyntax, cancellationToken) as IVariableDeclarationGroupOperation;
+            if (operation?.Declarations.Length != 1)
+            {
+                return default;
+            }
+
+            var localDeclaration = operation.Declarations[0];
+            if (localDeclaration.Declarators.Length != 1)
+            {
+                return default;
+            }
+
+            var declarator = localDeclaration.Declarators[0];
+            if (declarator.Initializer is null)
+            {
+                return default;
+            }
+
+            var localType = declarator.Symbol?.Type;
+            if (localType is null)
+            {
+                return default;
+            }
+
+            if (declarator.Initializer.Value is null ||
+                declarator.Initializer.Value.Kind == OperationKind.Invalid)
+            {
+                return default;
+            }
+
+            var isValidSelection = await CodeRefactoringHelpers.RefactoringSelectionIsValidAsync(
+                document,
+                selection,
+                node: declarationSyntax,
+                holes: ImmutableArray.Create(declarator.Initializer.Value.Syntax),
+                cancellationToken).ConfigureAwait(false);
```

Comment:
we should do the syntactic check before the more expensive semantic ones (i.e. IOperation).

Reply:
It uses IOperation in order to find the syntax for the hole, so it has to be last or else be specialized for every language.

Reply:
ah, pity.